### PR TITLE
PCHR-3368: Pass default status to leave request modal

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

This PR sets the default status to Approved when a manager wants to approve someone's leave request from the leave requests actions dropdown on the Manager Leave page.

## Technical Details

1. Leave popup service should be able to accept the default status to be set and pass it to the leave popup component:

```js
function openModal (/* ... */, defaultStatus) {
  // ...
  $modal.open({
    // ...
    resolve: {
      directiveOptions: function () {
        return {
          // ...
          defaultStatus: defaultStatus
```

2. Leave popup controller should accept the default status and set it to the UI as a default one

```js
function initStatus () {
  if (vm.defaultStatus) {
    vm.newStatusOnSave = vm.requestStatuses[sharedSettings.statusNames[vm.defaultStatus]].value;
  // ...
```

3. Finally, the actions dropdown needs to pass the default status depending on the action.

```js
var actionsToStatusesMap = {
  'approve': 'approved'

// ...

function openLeavePopupForAction (action) {
  return LeavePopup.openModal(vm.leaveRequest,
    // ...
    sharedSettings.statusNames[actionsToStatusesMap[action]]);
```

## Comments

The bug was found. The Leave Popup service was amended to not accept the Section param, but it's usage was not amended.

```js
// Before this PR
LeavePopup.openModal(
  vm.leaveRequest,
  vm.leaveRequest.request_type,
  vm.leaveRequest.contact_id,
  $rootScope.section === 'my-leave', // <<< should not be there
  true);

// After this PR
LeavePopup.openModal(vm.leaveRequest,
  vm.leaveRequest.request_type,
  vm.leaveRequest.contact_id,
  true,
  // notice, the unneeded param is gone
  // and a new legit param was added
  sharedSettings.statusNames[actionsToStatusesMap[action]]);
```

## Refactoring

This PR also refactors the `openModal` function of the `LeavePopup` service. This function now accepts an object with 5 properties instead of 5 arguments.

First of all, all places with this function were found:

```sh
grep -rl "openModal" ../uk.co.compucorp.civicrm.hrleaveandabsences
```
Then each file was manually checked and amended.

Also the file was deleted https://github.com/compucorp/civihr/pull/2706/commits/5d0aacbe3132e2deecc6e7551e06533221307bfb#diff-35d672bf94c4f53a21e0876c158b3c04

It was checked that this file is not used anywhere:
```
 grep -rl "staff-leave-calendar-month" ../
```
Also the last change on this file was made in August 2017, which probably means this file was forgotten to be deleted.

-----

✅Manual Tests - done
✅Jasmine Tests - augmented and passed
✅JS distributive files - included
⏹CSS distributive files - not needed
⏹Backstop tests - not needed
⏹PHP Unit Tests - not needed

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js [hunk 1]
```diff
@@ -274,6 +273,19 @@ define([
...
+    /**
+     * Opens a leave popup for a specific action
+     *
+     * @param {String} action
+     */
+    function openLeavePopupForAction (action) {
+      LeavePopup.openModal(vm.leaveRequest,
+        vm.leaveRequest.request_type,
+        vm.leaveRequest.contact_id,
+        true,
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: Is it always true? is there a place where it is false?
if not we dont need to send the param.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 2]
```diff
@@ -47,9 +47,10 @@ define([
      * @param {String} leaveType
      * @param {String} selectedContactId - Contact ID for the contact dropdown
      *                                     when the manager/admin is opening the request
-     * @param {Boolean} forceRecalculateBalanceChange optional
+     * @param {Boolean} [forceRecalculateBalanceChange]
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: Why is this inside square brackets?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js [hunk 3]
```diff
@@ -621,7 +643,6 @@
             var leaveRequest = LeaveRequestInstance.init(mockData.findBy('status_id', status));
 
             leaveRequest.contact_id = CRM.vars.leaveAndAbsences.contactId.toString();
-            role = 'manager';
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: Why this is removed?

**Comment by deb1990** [comment]
@igorpavlov The functionality works fine for admin, then do we need to add these new `defaultStatus ` and etc, how did it work before? and why cant we do the same for manager?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-actions.component.js]: No, it is not always true. This param forces the balance recalculation on modal open. It is used, yes, in the Leave Popup Service:

```js
function openModalByID (leaveRequestID) {
  return LeaveRequest.find(leaveRequestID)
    .then(function (leaveRequest) {
      return checkPermissionBeforeOpeningPopup(leaveRequest)
        .then(function (hasPermission) {
          if (hasPermission) {
            openModal(leaveRequest, leaveRequest.request_type, leaveRequest.contact_id);
          } else {
            notification.error('Error', 'You dont have permission to see this leave request');
          }
        });
    })
    .catch(function (errorMsg) {
      notification.error('Error', errorMsg);
    });
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: Optional param. http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js (hunk 3 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/request.controller.spec.js]: It is not removed, it is moved exactly at the place where it says "this is manager". 

https://github.com/compucorp/civihr/pull/2706/files#diff-e66ee11e23aa834592aec412809fbf34R596

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js [hunk 4]
```diff
@@ -633,10 +633,16 @@ define([
     }
 
     /**
-     * Initialises status
+     * Initialises status.
+     * If a default status is specified, then sets it.
+     * If not, sets the status to Approved if user is and admin
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: *if user is **an** admin*

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js [hunk 5]
```diff
@@ -47,9 +47,10 @@ define([
...
+     * @param {String} [defaultStatus] - "approved", "rejected" etc.
      */
-    function openModal (leaveRequest, leaveType, selectedContactId, forceRecalculateBalanceChange) {
+    function openModal (leaveRequest, leaveType, selectedContactId, forceRecalculateBalanceChange, defaultStatus) {
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: I don't enjoy the path this method is taking. It has too many parameters already and we are adding a new one.

Do you think it's possible, if time allows it, to fix this method? I think there are some things to investigate here:

* Do we really need the `leaveType` and `selectedContactId` parameters? Can't those be inferred by the `leaveRequest`'s object? (`leaveRequest.type_id` and `leaveRequest.contact_id`).
* Boolean parameters are really tricky. The problem is that they don't usually provide much context. If you see this implementation somewhere, what can you infer about the `true` parameter?
```js
leavePopup.openModal({id: 1, type_id: 3, contact_id: 5}, true);
```
I think we should use an "options" parameter instead which includes the `forceRecalculateBalanceChange` and `defaultStatus`:
```js
leavePopup.openModal(
  {id: 1, type_id: 3, contact_id: 5},
  {forceRecalculateBalanceChange: true, defaultStatus: 'approved'}
);

function openModal (leaveRequest, options) {
  // ...
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js (hunk 4 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request.controller.js]: Done

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js (hunk 5 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/services/leave-popup.service.js]: Done. I refactored the params.

**Pull request merged**